File: p3.py
#time: O(n)
#Space: O(1)
class Solution:
    def candy(self, ratings: List[int]) -> int:
        
        if ratings is None or len(ratings)==0:
            return 0
        
        candies = [1]*len(ratings)
        
        n = len(ratings)
        
        for i in range(1,n):
            if ratings[i-1] < ratings [i]:
                candies[i] = candies[i-1]+1
        
        for i in reversed(range(n-1)):
            if ratings[i] > ratings[i+1]:
                candies[i] = max(candies[i],candies[i+1]+1)
        
        
        return sum(candies)
            
        

File: p3.java
// Time Complexity :O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach

class Solution {
    public int candy(int[] ratings) {
        int[] candy = new int[ratings.length];
        int result = 0;

        //Make all the value of candy as 1
        for(int i=0; i<ratings.length; i++){
            candy[i] = 1;
        }
        
        //increment the value of candy if current rating is greater than previous rating
        for(int i=1; i<ratings.length; i++){
            if(ratings[i]>ratings[i-1]){
                candy[i] = candy[i-1] +1;
            }
        }
        
        //increment the value of candy if current rating is greater than previous rating
        result = candy[ratings.length-1];
        for(int i=ratings.length-2; i>=0; i--){
            if(ratings[i]>ratings[i+1]){
                candy[i] = Math.max(candy[i+1] +1, candy[i]);
                
            }
            result = result + candy[i];
        }
        
        return result;
    }
}